[Doc] Update KernelProgramCache.md with details on in-memory eviction

uditagarwal97 opened the pull request:
This PR updates KernelProgramCache.md by:
1. Making the doc in-sync with recent changes in in-memory cache (Like https://github.com/intel/llvm/pull/11351)
2. Adding details of in-memory cache eviction (https://github.com/intel/llvm/pull/16062)

Partially fixes: https://github.com/intel/llvm/issues/16017

maarquitos14 reviewed (commented):
UR is added in this new changes, but there are still plenty of mentions to PI throughout the document. Can we get them updated? PI does no longer exist.

Also, I like this kind of documents to be self-contained, so I don't like us pointing readers to EnvironmentVariables.md in Cache Configuration section. That's just a matter of style, so I will not block approval based on this, but I would like to have the environment variables listed in this document to make it easier for readers.

maarquitos14 commented:
@steffenlarsen since you reviewed #16062 too, you might want to take a look at this as well.

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -97,42 +97,27 @@ The cache is split into two levels:
-<a name="what-is-ksid">1</a>: Kernel set id is an ordinal number of the device
-binary image the kernel is contained in.
+The third map, called Fast Kernel Cache, is used as an optimization to reduce the
+number of lookups in the kernels map. It's key consists of the following components:
```
```suggestion
number of lookups in the kernels map. Its key consists of the following components:
```

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -408,10 +394,12 @@ LRU (least recently used) strategy both for in-memory and persistent cache.
-TODO: add detailed description of in-memory cache eviction mechanism.
+Eviction in in-memory cache is disabled by default but can be controlled by SYCL_IN_MEM_CACHE_EVICTION_THRESHOLD
+environment variable. The threshold is set in bytes and when the cache size exceeds the threshold the eviction process is initiated. The eviction process is based on LRU strategy. The cache is walked through and the least recently used items are deleted until the cache size is below the threshold.
+To implement eviction for in-memory cache efficiently, we store the programs in a linked-list, called eviction list. When the program is first added to the cache, it is also added to the back of the eviction list. When a program is fetched from cache, we move the program to the end of the eviction list. This way, we ensure that the programs at the beginning of the eviction list are always the lea ...
```
```suggestion
To implement eviction for in-memory cache efficiently, we store the programs in a linked-list, called the eviction list. When the program is first added to the cache, it is also added to the back of the eviction list. When a program is fetched from cache, we move the program to the end of the eviction list. This way, we ensure that the programs at the beginning of the eviction list are always the least recently used.
```

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -408,10 +394,12 @@ LRU (least recently used) strategy both for in-memory and persistent cache.
+Eviction in in-memory cache is disabled by default but can be controlled by SYCL_IN_MEM_CACHE_EVICTION_THRESHOLD
+environment variable. The threshold is set in bytes and when the cache size exceeds the threshold the eviction process is initiated. The eviction process is based on LRU strategy. The cache is walked through and the least recently used items are deleted until the cache size is below the threshold.
+To implement eviction for in-memory cache efficiently, we store the programs in a linked-list, called eviction list. When the program is first added to the cache, it is also added to the back of the eviction list. When a program is fetched from cache, we move the program to the end of the eviction list. This way, we ensure that the programs at the beginning of the eviction list are always the lea ...
+When adding a new program to cache, we check if the size of the program cache exceeds the threshold, if so, we iterate through the eviction list starting from the front and delete the programs until the cache size is below the threshold. When a program is deleted from the cache, we also evict its corresponding kernels from the kernel and fast kernel cache.
```
```suggestion
When adding a new program to cache, we check if the size of the program cache exceeds the threshold, if so, we iterate through the eviction list starting from the front and delete the programs until the cache size is below the threshold. When a program is deleted from the cache, we also evict its corresponding kernels from both of the kernel caches.
```
This wording seems a little less confusing, given kernel is mentioned in both names.

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -408,10 +394,12 @@ LRU (least recently used) strategy both for in-memory and persistent cache.
+To implement eviction for in-memory cache efficiently, we store the programs in a linked-list, called eviction list. When the program is first added to the cache, it is also added to the back of the eviction list. When a program is fetched from cache, we move the program to the end of the eviction list. This way, we ensure that the programs at the beginning of the eviction list are always the lea ...
+When adding a new program to cache, we check if the size of the program cache exceeds the threshold, if so, we iterate through the eviction list starting from the front and delete the programs until the cache size is below the threshold. When a program is deleted from the cache, we also evict its corresponding kernels from the kernel and fast kernel cache.
+
+***When the application run out-of-memory,*** either due to cache eviction being disabled or the cache eviction threshold being too high, we will evict all the items from program and kernel caches. This is done to prevent the application from crashing due to running out of memory.
```
```suggestion
***If the application runs out-of-memory,*** either due to cache eviction being disabled or the cache eviction threshold being too high, we will evict all the items from program and kernel caches.
```
The last part seems a little redundant to me. It makes sense to clean up the cache when we're running out of memory, given it's just a cache.

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -2,7 +2,7 @@
 ## Rationale behind caching
 
-During SYCL program execution SYCL runtime will create internal objects
+During SYCL program execution, SYCL runtime will create internal objects
```
```suggestion
During SYCL program execution, the SYCL runtime will create internal objects
```

steffenlarsen commented on sycl/doc/design/KernelProgramCache.md:
````diff
@@ -65,7 +65,7 @@ examples below illustrate scenarios where such optimization is possible.
 ```
 
-In both cases SYCL runtime will need to build the program and kernels multiple
+In both cases, SYCL runtime will need to build the program and kernels multiple
````
```suggestion
In both cases, the SYCL runtime will need to build the program and kernels multiple
```

steffenlarsen reviewed (commented):
A couple of comments, mostly editorial and for consistency.

uditagarwal97 commented:
> UR is added in this new changes, but there are still plenty of mentions to PI throughout the document. Can we get them updated? PI does no longer exist.
> 
> Also, I like this kind of documents to be self-contained, so I don't like us pointing readers to EnvironmentVariables.md in Cache Configuration section. That's just a matter of style, so I will not block approval based on this, but I would like to have the environment variables listed in this document to make it easier for readers.

In e24d975ff60523e6542d435e807b4bd9531c2c61, I've removed remaining PI references and added cache-related environment variables in this doc itself.

steffenlarsen reviewed (approved):
LGTM!

maarquitos14 commented on sycl/doc/design/KernelProgramCache.md:
```diff
@@ -177,20 +162,33 @@ values for `info::platform::name`, `info::device::name`,
+| Environment variable | Values | Description |
+| -------------------- | ------ | ----------- |
+| `SYCL_CACHE_DIR` | Path | Path to persistent cache root directory. Default values are `%AppData%\libsycl_cache` for Windows and `$XDG_CACHE_HOME/libsycl_cache` on Linux, if `XDG_CACHE_HOME` is not set then `$HOME/.cache/libsycl_cache`. When none of the environment variables are set SYCL persistent cache is disabled. |
+| `SYCL_CACHE_PERSISTENT` | Integer | Controls persistent device compiled code cache. Turns it on if set to '1' and turns it off if set to '0'. When cache is enabled SYCL runtime will try to cache and reuse JIT-compiled binaries. Default is off. |
```
```suggestion
| `SYCL_CACHE_PERSISTENT` | '1' or '0' | Controls persistent device compiled code cache. Turns it on if set to '1' and turns it off if set to '0'. When cache is enabled SYCL runtime will try to cache and reuse JIT-compiled binaries. Default is off. |
```
For consistency with the next variable. Or either 'Integer', I don't have a strong opinion, but would like both to be the same.

maarquitos14 reviewed (approved):
Just a nit, otherwise LGTM.

uditagarwal97 commented:
@intel/llvm-gatekeepers the PR is ready to be merged.
